lib/Dataflow_analysis.ml

```diff
@@ -0,0 +1,800 @@
+open Core_kernel
+open Mir
+
+(***********************************)
+(* Basic datatypes                 *)
+(***********************************)
+
+(**
+   A label is a unique identifier for a node in the dataflow/dependency graph, and
+   usually corresponds to one node in the Mir.
+*)
+type label = int [@@deriving sexp]
+
+(**
+   A 'reaching dependency' (or reaching_dep or RD) statement (v, l) says that the variable
+   v could have been affected at the label l.
+*)
+type reaching_dep = expr * int [@@deriving sexp, hash, compare]
```

Comment:
Yep, this should be expr * label, and I think the `deriving hash` might actually be unnecessary.

I think we could use arrays instead of LabelMap, I just didn't foresee it being a performance issue and Map was convenient.

Yes, in the current implementation `expr` is just used to represent a Var, but in the future it'll also include indexing information. I think it would actually be best to split Var and Indexed out of Mir.expr into its own type, but I didn't want to make that change in Mir without discussing it.

Reply:
Feel free to ping me here on github, via email (my name at gmail), or even WhatsApp if you want to bounce ideas off me quickly. Also happy to hop on quick impromptu hangouts as needed.

I think the MIR will always need a variant in `expr` that represents a `Var`, right? What I've done in the rest of the backend so far has been to represent standalone Vars as the underlying `string`s, since I'm not a huge fan of type aliases so far. In that sense, a Var is already split out :P But Matthijs represents an identifier as a more complex data structure (that also has more information, as well as type aliasing `string`). 

My preference against type aliasing so far comes mostly because 1) you need to redefine a bunch of functions to work on the new alias, yet 2) it doesn't actually seem to prevent passing a normal `int` in for a `label` or vice versa, and 3) my editor deals poorly with type aliases such that throughout the whole module even in unrelated functions, `int` will display as `label`, which is confusing :P 

Is the basic idea here in a `reaching_dep` that you need to have an LHS, which is either a Var or some indexed piece thereof? Indices are already a separate type, though I'm seeing now that they take an `expr` instead of just a `string`. I think this is the same issue and should probably be resolved the same way as above (oops).

Reply:
Great, I'll send you a note if I run into something like that in the future.

The MIR will definitely need Var as a variant, but it could still be in a different type with Indexed, something like
`type val = Var .. | Indexed ...`
`type expr = ... | Val of val`
It would be more annoying to use in some places, but would make more sense as an abstraction in this section. Or I can define it myself here, and write a translation from expr -> val option.

Regarding type aliases: 1. can't you still use functions that work on int for labels? 2. True, I think you'd have to give it a constructor to make it safe 3. I've also run into this problem, it's annoying.
I'm coming from Haskell, where type aliases only make type signatures more readable, and there's a different option called a newtype that has a zero-cost constructor and prevents you from passing it in place of the underlying type. I'm perfectly happy to avoid them in OCaml where they don't seem to be quite as pain-free.

I agree, there should be a type for Val | Indexed split off.